gslib/tests/test_hmac.py

```diff
@@ -112,24 +123,44 @@ def CreateHelper(self, service_account):
 
   def test_malformed_commands(self):
     params = [
-        ('hmac create', 'requires a service account'),
-        ('hmac create -u email', 'requires a service account'),
-        ('hmac create -p proj', 'requires a service account'),
-        ('hmac delete', 'requires an Access ID'),
-        ('hmac delete -p proj', 'requires an Access ID'),
-        ('hmac get', 'requires an Access ID'),
-        ('hmac get -p proj', 'requires an Access ID'),
-        ('hmac list account1', 'unexpected arguments'),
-        ('hmac update keyname', 'state flag must be supplied'),
-        ('hmac update -s KENTUCKY', 'state flag value must be one of'),
-        ('hmac update -s INACTIVE', 'requires an Access ID'),
-        ('hmac update -s INACTIVE -p proj', 'requires an Access ID'),
+        ('hmac create', 'requires a service account',
+         'argument SERVICE_ACCOUNT: Must be specified'),
+        ('hmac create -u email', 'requires a service account', ''),
+        ('hmac create -p proj', 'requires a service account',
+         'argument SERVICE_ACCOUNT: Must be specified'),
+        ('hmac delete', 'requires an Access ID',
+         'argument ACCESS_ID: Must be specified'),
+        ('hmac delete -p proj', 'requires an Access ID',
+         'argument ACCESS_ID: Must be specified'),
+        ('hmac get', 'requires an Access ID',
+         'argument ACCESS_ID: Must be specified'),
+        ('hmac get -p proj', 'requires an Access ID',
+         'argument ACCESS_ID: Must be specified'),
+        ('hmac list account1', 'unexpected arguments',
+         'unrecognized arguments'),
+        ('hmac update keyname', 'state flag must be supplied',
+         'Exactly one of (--activate | --deactivate) must be specified.'),
+        ('hmac update -s KENTUCKY', 'state flag value must be one of', ''),
+        ('hmac update -s INACTIVE', 'requires an Access ID',
+         'argument ACCESS_ID: Must be specified'),
+        ('hmac update -s INACTIVE -p proj', 'requires an Access ID',
+         'argument ACCESS_ID: Must be specified'),
     ]
-    for command, error_substr in params:
+    for command, gsutil_error_substr, gcloud_error_substr in params:
+      flag_cannot_be_mapped_to_shim = not gcloud_error_substr
+      if self._use_gcloud_storage and flag_cannot_be_mapped_to_shim:
```

Comment:
`if` statements in a unit test are a strong indicator there should be two separate tests.

We allow this pattern:
```
if self._use_gcloud_storage:
  self.assertIn(gcloud_error_substr, stderr)
else:
  self.assertIn(gsutil_error_substr, stderr)
```

However, with the `continue` statement, this test has gotten too complex. We want tests to tell us what to debug and not to debug tests.

Reply:
I split this test into 2 tests, one called `test_malformed_commands` and the another one called `test_malformed_commands_that_cannot_be_translated_using_the_shim`, the first is for all malformed commands cases that can be tested using the shim, the another test is for all malformed command cases that cannot be translated by the shim and the test is skipped if it is running with the shim enabled.